pkg/cmd/get/token/exec.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		return err
+	}
+	if !installed {
+		return fmt.Errorf("this is not a hub")
+	}
+	return err
+}
+
+func (o *Options) run() error {
+	output := make([]string, 0)
+	reader := scenario.GetScenarioResourcesReader()
+
+	kubeClient, err := o.ClusteradmFlags.KubectlFactory.KubernetesClientSet()
+	if err != nil {
+		return err
+	}
+	dynamicClient, err := o.ClusteradmFlags.KubectlFactory.DynamicClient()
+	if err != nil {
+		return err
+	}
+
+	restConfig, err := o.ClusteradmFlags.KubectlFactory.ToRESTConfig()
+	if err != nil {
+		return err
+	}
+
+	apiExtensionsClient, err := apiextensionsclient.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	clientHolder := resourceapply.NewClientHolder().
+		WithAPIExtensionsClient(apiExtensionsClient).
+		WithKubernetes(kubeClient).
+		WithDynamicClient(dynamicClient)
+
+	token, err := getToken(kubeClient)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		files := []string{
+			"init/namespace.yaml",
+		}
+		if o.useBootstrapToken {
+			files = append(files,
+				"init/bootstrap-token-secret.yaml",
+				"init/bootstrap_cluster_role.yaml",
+				"init/bootstrap_cluster_role_binding.yaml",
+			)
+		} else {
+			files = append(files,
+				"init/bootstrap_sa.yaml",
+				"init/bootstrap_cluster_role.yaml",
+				"init/bootstrap_sa_cluster_role_binding.yaml",
+			)
+		}
+		out, err := apply.ApplyDirectly(clientHolder, reader, o.values, o.ClusteradmFlags.DryRun, "", files...)
+		if err != nil {
+			return err
+		}
+		output = append(output, out...)
+		if !o.ClusteradmFlags.DryRun {
+			//Make sure that the sa token is ready
+			if !o.useBootstrapToken {
+				b := retry.DefaultBackoff
+				b.Duration = 100 * time.Millisecond
+				_, err = waitForBootstrapSecret(kubeClient, b)
+				if err != nil {
+					return err
+				}
+			}
+			token, err = getToken(kubeClient)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		//Update the cluster role
+		files := []string{
+			"init/bootstrap_cluster_role.yaml",
+		}
+		out, err := apply.ApplyDirectly(clientHolder, reader, o.values, o.ClusteradmFlags.DryRun, "", files...)
+		if err != nil {
+			return err
+		}
+		output = append(output, out...)
+	}
+	fmt.Printf("token=%s\n", token)
+
+	fmt.Printf("please log on spoke and run:\n%s join --hub-token %s --hub-apiserver %s --cluster-name <cluster_name>\n",
+		helpers.GetExampleHeader(),
+		token,
+		restConfig.Host,
+	)
+
+	return apply.WriteOutput(o.outputFile, output)
+}
+
+func waitForBootstrapSecret(kubeClient kubernetes.Interface, b wait.Backoff) (secret *corev1.Secret, err error) {
+	err = retry.OnError(b, func(err error) bool {
+		return err != nil
+	}, func() error {
+		secret, err = helpers.GetBootstrapSecret(kubeClient)
+		return err
+	})
+	return
+}
+
+func getToken(kubeClient kubernetes.Interface) (string, error) {
+	var bootstrapSecret *corev1.Secret
+	saSecret, err := helpers.GetBootstrapSecret(kubeClient)
```

Comment:
this func name GetBootstrapSecret is confusing, i would change to GetBootstrapSecretFromSA

Reply:
Done